### PR TITLE
feat(nextjs): Support rendering ClerkProvider in client components

### PR DESCRIPTION
**Pull request by BRKalow**
## Description

Allow `ClerkProvider` to be rendered in client components in App Router. This is accomplished by introducing a "compat" layer that uses `next/compat/router` to detect the presence of the pages router. If detected, we render the pages provider, otherwise we render the app router provider.

fixes JS-573

## Checklist

- [x] `npm test` runs as expected.
- [x] `npm run build` runs as expected.
- [ ] (If applicable) [JSDoc comments](https://jsdoc.app/about-getting-started.html) have been added or updated for any package exports
- [ ] (If applicable) [Documentation](https://github.com/clerkinc/clerk-docs) has been updated

## Type of change

- [ ] 🐛 Bug fix
- [x] 🌟 New feature
- [ ] 🔨 Breaking change
- [ ] 📖 Refactoring / dependency upgrade / documentation
- [ ] other:

## Packages affected

- [ ] `@clerk/clerk-js`
- [ ] `@clerk/clerk-react`
- [x] `@clerk/nextjs`
- [ ] `@clerk/remix`
- [ ] `@clerk/types`
- [ ] `@clerk/themes`
- [ ] `@clerk/localizations`
- [ ] `@clerk/clerk-expo`
- [ ] `@clerk/backend`
- [ ] `@clerk/clerk-sdk-node`
- [ ] `@clerk/shared`
- [ ] `@clerk/fastify`
- [ ] `@clerk/chrome-extension`
- [ ] `gatsby-plugin-clerk`
- [ ] `build/tooling/chore`

### CODE REVIEW

**Comment by BRKalow** [comment]
!snapshot

**Comment by clerk-cookie** [comment]
Hey @BRKalow - the snapshot version command generated the following package versions:
| Package | Version |
| --- | --- |
| @clerk/backend | 0.30.3-snapshot.e42e882 |
| @clerk/chrome-extension | 0.4.6-snapshot.e42e882 |
| @clerk/clerk-js | 4.60.1-snapshot.e42e882 |
| @clerk/clerk-expo | 0.19.8-snapshot.e42e882 |
| @clerk/fastify | 0.6.13-snapshot.e42e882 |
| gatsby-plugin-clerk | 4.4.14-snapshot.e42e882 |
| @clerk/nextjs | 4.25.3-snapshot.e42e882 |
| @clerk/clerk-react | 4.26.3-snapshot.e42e882 |
| @clerk/remix | 3.0.5-snapshot.e42e882 |
| @clerk/clerk-sdk-node | 4.12.12-snapshot.e42e882 |
| @clerk/shared | 0.24.3-snapshot.e42e882 |

Tip: use the snippet copy button below to quickly install the required packages.
```sh
# @clerk/backend
npm i @clerk/backend@0.30.3-snapshot.e42e882
```
```sh
# @clerk/chrome-extension
npm i @clerk/chrome-extension@0.4.6-snapshot.e42e882
```
```sh
# @clerk/clerk-js
npm i @clerk/clerk-js@4.60.1-snapshot.e42e882
```
```sh
# @clerk/clerk-expo
npm i @clerk/clerk-expo@0.19.8-snapshot.e42e882
```
```sh
# @clerk/fastify
npm i @clerk/fastify@0.6.13-snapshot.e42e882
```
```sh
# gatsby-plugin-clerk
npm i gatsby-plugin-clerk@4.4.14-snapshot.e42e882
```
```sh
# @clerk/nextjs
npm i @clerk/nextjs@4.25.3-snapshot.e42e882
```
```sh
# @clerk/clerk-react
npm i @clerk/clerk-react@4.26.3-snapshot.e42e882
```
```sh
# @clerk/remix
npm i @clerk/remix@3.0.5-snapshot.e42e882
```
```sh
# @clerk/clerk-sdk-node
npm i @clerk/clerk-sdk-node@4.12.12-snapshot.e42e882
```
```sh
# @clerk/shared
npm i @clerk/shared@0.24.3-snapshot.e42e882
```

File: packages/nextjs/src/client-boundary/ClerkProvider.tsx [hunk 1]
```diff
@@ -0,0 +1,19 @@
+'use client';
+
+import { useRouter } from 'next/compat/router';
```
Comment (BRKalow) [inline packages/nextjs/src/client-boundary/ClerkProvider.tsx:3]: More details from the exported hook here: https://github.com/vercel/next.js/blob/canary/packages/next/src/client/compat/router.ts

File: packages/nextjs/src/client-boundary/ClerkProvider.tsx [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+import { ClerkProvider as PageClerkProvider } from '../pages/ClerkProvider';
+import { type NextClerkProviderProps } from '../types';
+
+/**
+ * This is a compatibility layer to support a single ClerkProvider component in both the app and pages routers.
+ */
+export function ClerkProvider(props: NextClerkProviderProps) {
+  const router = useRouter();
+
+  const Provider = router ? PageClerkProvider : ClientClerkProvider;
```
Comment (nikosdouvlis) [inline packages/nextjs/src/client-boundary/ClerkProvider.tsx:16]: The solution here makes perfect sense to me :) 
Do you think we could merge the two client providers into one? I took a quick look and they look pretty similar.

**Review by nikosdouvlis — APPROVED** [review]

File: packages/nextjs/src/client-boundary/ClerkProvider.tsx (hunk 2 above)
Comment (BRKalow) [reply packages/nextjs/src/client-boundary/ClerkProvider.tsx:16]: @nikosdouvlis possibly, the difficulty is they rely on the two different router hooks, which is why we need this layer. We could abstract the dependency on the router into a prop and use this layer to compute the prop to pass down to single provider.

File: packages/nextjs/src/client-boundary/ClerkProvider.tsx (hunk 1 above)
Comment (LekoArts) [reply packages/nextjs/src/client-boundary/ClerkProvider.tsx:3]: Is this a stable export we can rely on being available?

Or are we reaching into private APIs here?

Comment (BRKalow) [reply packages/nextjs/src/client-boundary/ClerkProvider.tsx:3]: I believe it is stable and reliable, based on PR: https://github.com/vercel/next.js/pull/42502.

File: integration/tests/next-build.test.ts [hunk 3]
```diff
@@ -0,0 +1,72 @@
+import { expect, test } from '@playwright/test';
+
+import type { Application } from '../models/application';
+import { appConfigs } from '../presets';
+
+test.describe('next build @nextjs', () => {
```
Comment (BRKalow) [inline integration/tests/next-build.test.ts:6]: @nikosdouvlis Any feedback on the test? 🙏

File: package.json [hunk 4]
```diff
@@ -69,7 +69,7 @@
     "test:ci": "FORCE_COLOR=1 turbo test --concurrency=${TURBO_CONCURRENCY:-80%}",
     "test:integration:base": "DEBUG=1 npx playwright test --config integration/playwright.config.ts",
     "test:integration:generic": "E2E_APP_ID=react.vite.* npm run test:integration:base -- --grep @generic",
-    "test:integration:nextjs": "E2E_APP_ID=next.appRouter.withEmailCodes npm run test:integration:base -- --grep @generic",
+    "test:integration:nextjs": "E2E_APP_ID=next.appRouter.withEmailCodes npm run test:integration:base -- --grep \"@generic|@nextjs\"",
```
Comment (BRKalow) [inline package.json:72]: As discussed in the initial RFC, I added a `@nextjs` tag as well.

File: integration/models/application.ts [hunk 5]
```diff
@@ -70,7 +71,16 @@ export const application = (config: ApplicationConfig, appDirPath: string, appDi
     },
     build: async () => {
       const log = logger.child({ prefix: 'build' }).info;
-      await run(scripts.build, { cwd: appDirPath, log });
+      await run(scripts.build, {
+        cwd: appDirPath,
+        log: (msg: string) => {
+          buildOutput += `\n${msg}`;
+          log(msg);
+        },
```
Comment (BRKalow) [inline integration/models/application.ts:79]: Captures and makes the build output available by way of `app.buildOutput`. This allows us to assert on the output.

**Review by dimkl — APPROVED** [review]

**Comment by clerk-cookie** [comment]
This PR has been automatically locked since there has not been any recent activity after it was closed. Please open a new issue for related bugs.

**Pull request merged**